### PR TITLE
Refactor rewriting of Spark yarn Client into cdap-spark-core module

### PR DESCRIPTION
**Pull request by chtyim**

### CODE REVIEW

**Comment by chtyim** [comment]
Build running in http://builds.cask.co/browse/CDAP-DUT3940

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java [hunk 1]
```diff
@@ -68,9 +69,11 @@ public ProgramRunner createProgramRunner(ProgramType type, Mode mode, Injector i
    */
   private ProgramRunner createSparkProgramRunner(Injector injector, String programRunnerClassName) {
     try {
-      CConfiguration cConf = injector.getInstance(CConfiguration.class);
-      final SparkRunnerClassLoader classLoader = new SparkRunnerClassLoader(getClassLoaderURLs(cConf),
-                                                                            getClass().getClassLoader());
+      boolean rewriteYarnClient = injector.getInstance(CConfiguration.class)
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: when should this property should be set enabled? It depends on the spark version, right? So how do we communicate that to users?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java [hunk 2]
```diff
@@ -298,4 +314,96 @@ public MethodVisitor visitMethod(int access, String name, String desc, String si
       return null;
     }
   }
+
+  /**
+   * Defines the org.apache.spark.deploy.yarn.Client class with rewriting of the createConfArchive method to
+   * workaround the SPARK-13441 bug.
+   */
+  private Class<?> defineClient(String name, InputStream createConfArchive) throws IOException, ClassNotFoundException {
+    // We only need to rewrite if listing either HADOOP_CONF_DIR or YARN_CONF_DIR return null.
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: why is it so?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java (hunk 1 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: By default, it is always enabled. Without YARN-4727 being fixed, on CM cluster, we still need this even the Spark bug is patched.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java (hunk 2 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: That's what the SPARK-13441 is.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java [hunk 3]
```diff
@@ -298,4 +314,96 @@ public MethodVisitor visitMethod(int access, String name, String desc, String si
...
+
+        // push three constants to the stack
+        mg.visitLdcInsn(SPARK_CONF_FILE);
+        mg.visitLdcInsn(LOCALIZED_CONF_DIR);
+        mg.visitLdcInsn(LOCALIZED_CONF_DIR_ZIP);
+
+        // call SparkRuntimeUtils.createConfArchive, return a File and leave it in stack
+        Type stringType = Type.getType(String.class);
+        mg.invokeStatic(Type.getType(SparkRuntimeUtils.class),
+                        new Method("createConfArchive", Type.getType(File.class),
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: Any reason for not using ``Methods.getMethod`` here (that was how it was earlier)

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java [hunk 4]
```diff
@@ -298,4 +314,96 @@ public MethodVisitor visitMethod(int access, String name, String desc, String si
...
+                                   new Type[] { SPARK_CONF_TYPE, stringType, stringType, stringType}));
+        if (isReturnFile) {
+          // Spark 1.5+ return type is File, hence just return the File from the stack
+          mg.returnValue();
+          mg.endMethod();
+        } else {
+          // Spark 1.4 return type is Option<File>
+          // return Option.apply(<file from stack>);
+          // where the file is actually just popped from the stack
+          mg.invokeStatic(optionType, new Method("apply", optionType, new Type[] { Type.getType(Object.class) }));
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: ditto

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java [hunk 5]
```diff
@@ -298,4 +314,96 @@ public MethodVisitor visitMethod(int access, String name, String desc, String si
...
+
+        // Only rewrite the createConfArchive method
+        if (!"createConfArchive".equals(name)) {
+          return mv;
+        }
+
+        // Check if it's a recognizable return type.
+        // Spark 1.5+ return type is File
+        boolean isReturnFile = Type.getReturnType(desc).equals(Type.getType(File.class));
+        Type optionType = Type.getObjectType("scala/Option");
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: Don't want direct dependencies on scala classes in this class?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java (hunk 2 above)
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: oh the NPE! Got it :-)

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java (hunk 1 above)
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: Right - I meant how do we communicate that to users? Should we update the property description to include more details?

**Comment by bdmogal** [comment]
Just a couple of questions from me. Otherwise the refactoring LGTM

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java (hunk 3 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: For this case, no particular reason. However, in this class, we don't use `Methods.getMethod` in other cases due to we cannot refer to Spark/Scala classes with ".class". Also, I would like to minimize dependencies on cdap core classes so that in near future, we can provide better classloader protection for runtime extensions.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java (hunk 4 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: That is the ditto?

Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: Oh yeah. In this case, we cannot use "Option.class".

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java (hunk 1 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: This is the current description: 
Specify whether to rewrite the yarn.Client class in Spark to work around the issue SPARK-13441 in CDM clusters.

Do we need to update it?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java (hunk 5 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: Right, due to future classloader isolation requirement.

**Comment by chtyim** [comment]
Build: http://builds.cask.co/browse/CDAP-DUT3940

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java (hunk 1 above)
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: Do you think it would be good to tell users when it can be set to false (in future)?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java (hunk 3 above)
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: 👍

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java (hunk 4 above)
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: I meant why the change to not use ``Methods.getMethod``, but you answered that already

**Comment by bdmogal** [comment]
@chtyim other than https://github.com/caskdata/cdap/pull/5530#discussion_r59657850, LGTM!

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java (hunk 1 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: It would be a bit difficult to describe since it is a combination of three bugs/behaviors that require the rewrite. I would say leave it on as long as it is working.

Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: Besides, there are logic in place to decide if rewrite is actually needed.

Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: 👍

**Comment by chtyim** [comment]
Build passed
